packages/contracts/test/utils/test_with_reference.ts

```diff
@@ -0,0 +1,98 @@
+import * as chai from 'chai';
+import * as _ from 'lodash';
+
+import { chaiSetup } from './chai_setup';
+
+chaiSetup.configure();
+const expect = chai.expect;
+
+export async function testWithReferenceFuncAsync<P0, P1, R>(
+    referenceFunc: (p0: P0, p1: P1) => Promise<R>,
+    testFunc: (p0: P0, p1: P1) => Promise<R>,
+    values: [P0, P1],
+): Promise<void>;
+export async function testWithReferenceFuncAsync<P0, P1, P2, R>(
+    referenceFunc: (p0: P0, p1: P1, p2: P2) => Promise<R>,
+    testFunc: (p0: P0, p1: P1, p2: P2) => Promise<R>,
+    values: [P0, P1, P2],
+): Promise<void>;
+export async function testWithReferenceFuncAsync<P0, P1, P2, P3, R>(
+    referenceFunc: (p0: P0, p1: P1, p2: P2, p3: P3) => Promise<R>,
+    testFunc: (p0: P0, p1: P1, p2: P2, p3: P3) => Promise<R>,
+    values: [P0, P1, P2, P3],
+): Promise<void>;
+export async function testWithReferenceFuncAsync<P0, P1, P2, P3, P4, R>(
+    referenceFunc: (p0: P0, p1: P1, p2: P2, p3: P3, p4: P4) => Promise<R>,
+    testFunc: (p0: P0, p1: P1, p2: P2, p3: P3, p4: P4) => Promise<R>,
+    values: [P0, P1, P2, P3, P4],
+): Promise<void>;
```

Comment:
Why are these types defined in two places? Let's declare them in one place and re-export them in the other.

Reply:
I'm not sure what you mean exactly. There is a subtle differences between the type definitions for `testWithReferenceFuncAsync` and `testCombinatoriallyWithReferenceFuncAsync`. Namely, the combinatorial one expects an array of arrays of values instead of just an array of values.

The only thing I can think of to reduce code duplication here in the type definitions is to define types like this:

```typescript
type asyncFunc2<P0, P1, R> = (p0: P0, p1: P1) => Promise<R>;
type asyncFunc3<P0, P1, P2, R> = (p0: P0, p1: P1, p2: P2) => Promise<R>;
// ...
```

And then use them like this:

```typescript
export async function testWithReferenceFuncAsync<P0, P1, R>(
    referenceFunc: asyncFunc2<P0, P1, R>,
    testFunc: asyncFunc2<P0, P1, R>,
    values: [P0, P1],
): Promise<void>;
export async function testWithReferenceFuncAsync<P0, P1, P2, R>(
    referenceFunc: asyncFunc3<P0, P1, P2, R>,
    testFunc: asyncFunc3<P0, P1, P2, R>,
    values: [P0, P1, P2],
): Promise<void>;
// ...
```

But IMO this is less readable because it requires readers to then look at the definition for `asyncFunc2`, `asyncFunc3`, etc. And it doesn't really make our code any shorter either.

I did go ahead rename the `values` argument in `testCombinatoriallyWithReferenceFuncAsync` to `allValues` to make the difference more obvious. Do you have any other ideas? Maybe there's something I'm missing.

Reply:
Makes sense. Nope, think this works.